### PR TITLE
[Quant][fx][bc-breaking] Replace qconfig_dict with a config object (OLD, not merged)

### PR DESCRIPTION
**Pull request by andrewor14**
**Summary:** Previously, FX graph mode quantization configurations
were specified through a dictionary of qconfigs. However, this
API was not in line with other core APIs in PyTorch. This commit
replaces this dictionary with a config object that users will
create and pass to prepare and convert. This leads to better
type safety and better user experience in notebook settings
due to improved auto completion.

The new API is as follows:

```
from torch.ao.quantization import QConfigMapping
from torch.ao.quantization.quantize_fx import prepare_fx

qconfig_mapping = QConfigMapping()
    .set_global(qconfig)
    .set_object_type(torch.nn.Linear, qconfig)
    .set_module_name_regex("foo.*bar", qconfig)
    .set_module_name("mod", qconfig)

prepare_fx(model, qconfig_mapping)
```

For backwards compatibility, `prepare_fx`, `prepare_qat_fx`,
and `convert_fx` will continue to accept qconfig_dicts, which
will be converted to QuantizationConfigs internally.

Note that this commit does not modify existing tests to use the
new API; they will continue to pass in qconfig_dict as before,
which still works but triggers a deprecation warning. This will
be handled in a future commit.

**Test Plan:**
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

**Reviewers:** jerryzh168, vkuzo, HDCharles

**Subscribers:** jerryzh168, vkuzo, HDCharles

Differential Revision: [D36597438](https://our.internmc.facebook.com/intern/diff/D36597438)

### CODE REVIEW

File: torch/ao/quantization/quantize_fx.py [hunk 1]
```diff
@@ -196,12 +198,27 @@ def _prepare_fx(
...
+    if equalization_qconfig_container is None:
+        equalization_qconfig_container = QConfigContainer()
 
-    check_is_valid_qconfig_dict(qconfig_dict)
     check_is_valid_prepare_custom_config_dict(prepare_custom_config_dict)
-    check_is_valid_qconfig_dict(equalization_qconfig_dict)
+
+    if isinstance(qconfig_container, Dict):
+        warnings.warn(
+            "Passing a QConfig dictionary to prepare is deprecated. "
```
Comment (vkuzo) [inline torch/ao/quantization/quantize_fx.py:208]: we can also mention that this will be removed in a future PyTorch version

File: torch/ao/quantization/quantize_fx.py [hunk 2]
```diff
@@ -196,12 +198,27 @@ def _prepare_fx(
     """
     if prepare_custom_config_dict is None:
         prepare_custom_config_dict = {}
-    if equalization_qconfig_dict is None:
-        equalization_qconfig_dict = {}
+    if equalization_quantization_config is None:
```
Comment (jerryzh168) [inline torch/ao/quantization/quantize_fx.py:201]: I think equalization_quantization_config is a bit weird, maybe just `equalization_config`?

**Comment by jerryzh168** [comment]
Please hold on the refactor for internal files for now to reduce merge conflict, the example_inputs change will be touching these things as well

**Comment by andrewor14** [comment]
Ok, how about I wait for your example_inputs change to be landed first? Then I'll deal with the conflicts in this PR. Meanwhile I still need to fix the remaining tests and fill in the comments.

File: test/allowlist_for_publicAPI.json [hunk 3]
```diff
@@ -130,14 +130,15 @@
     "Node",
     "Optional",
     "QConfigAny",
+    "QConfigMapping",
```
Comment (albanD) [inline test/allowlist_for_publicAPI.json:133]: No new entry should be added to this list!
Make sure to properly setup your module to reflect what is public and what is not public.

**Review by albanD — CHANGES_REQUESTED** [review]

File: test/allowlist_for_publicAPI.json (hunk 3 above)
Comment (jerryzh168) [reply test/allowlist_for_publicAPI.json:133]: I remember last time I tried there were some problems with this test, wondering if this is fixed now or not

Comment (albanD) [reply test/allowlist_for_publicAPI.json:133]: Yes, this test has been working fine for the past month now.

**Comment by andrewor14** [comment]
Hi @albanD, I removed the changes in `test/allowlist_for_publicAPI.json`. Please have another look.

File: test/allowlist_for_publicAPI.json (hunk 3 above)
Comment (andrewor14) [reply test/allowlist_for_publicAPI.json]: Ok, I removed the changes in this file. Please have another look @albanD.

File: torch/ao/quantization/fx/convert.py [hunk 4]
```diff
@@ -59,6 +60,27 @@
...
+    "duplicate_quantize_dynamic_node",
+    "get_module_path_and_prefix",
+    "has_none_qconfig",
+    "insert_dequantize_node",
+    "maybe_get_observer_for_node",
+    "maybe_recursive_remove_dequantize",
+    "remove_extra_dequantize",
+    "remove_quant_dequant_pairs",
+    "restore_state",
+    "run_weight_observers",
```
Comment (andrewor14) [inline torch/ao/quantization/fx/convert.py:80]: In this PR I just wanted to preserve the existing behavior for unrelated functions. We should further restrict the visibility here and other places but let's leave that for a future PR.

File: torch/ao/quantization/qconfig_mapping_utils.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
+from collections import OrderedDict
```
Comment (andrewor14) [inline torch/ao/quantization/qconfig_mapping_utils.py:1]: This file was renamed from `qconfig_dict_utils.py`

File: test/quantization/fx/test_quantize_fx.py [hunk 6]
```diff
@@ -1821,21 +1845,180 @@ def forward(self, x):
                 # should not crash as in https://github.com/pytorch/pytorch/issues/75825
                 prepare_fx(m, qconfig_dict)
 
-    def test_qconfig_dict_validity(self):
-        r"""
-        Verifies that if a user passes an invalid key or makes a typo when
-        constructing a qconfig_dict, an error will be thrown and users will be
-        notified of what keys are supported.
+    def test_qconfig_mapping_set_global(self):
+        qconfig = QConfig(123, "global")
```
Comment (vkuzo) [inline test/quantization/fx/test_quantize_fx.py:1849]: this may work now because QConfig does not typecheck, but at some point this may break.  It might make sense to just use a real constructor.

File: torch/ao/quantization/qconfig_mapping.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+@dataclass
+class QConfigModuleNameObjectTypeOrderEntry:
+    module_name: str
+    object_type: Callable
+    index: int
+    qconfig: QConfigAny
+
+
+class QConfigMapping:
```
Comment (vkuzo) [inline torch/ao/quantization/qconfig_mapping.py:56]: it would be nice to have rendered documentation for this class, can be in a separate PR

**Review by vkuzo — APPROVED** [review]
lg as long as CI passes, thanks for working on this

File: torch/ao/quantization/quantize_fx.py [hunk 8]
```diff
@@ -339,57 +356,33 @@ def fuse_fx(
...
-            # qconfig == None means fusion and quantization should be skipped for anything
-            # matching the rule
-          }
+      * `qconfig_mapping`: mapping from model ops to qconfigs::
+
+          qconfig_mapping = QConfigMapping()
+              # default_qconfig
+              .set_global(global_qconfig)
+              # match on module type, function, or method name
+              .set_object_type(torch.nn.Linear, qconfig1)
```
[the remark this answers is not in the document]
Comment (andrewor14) [reply torch/ao/quantization/quantize_fx.py:375]: Let me verify

File: torch/ao/quantization/fx/convert.py [hunk 9]
```diff
@@ -513,7 +535,7 @@ def convert(
         convert_custom_config_dict: Dict[str, Any] = None,
         is_standalone_module: bool = False,
         _remove_qconfig_flag: bool = True,
-        convert_qconfig_dict: Dict[str, Any] = None,
+        qconfig_mapping: QConfigMapping = None,
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/convert.py:538]: nit: feels like this should be `Optional[QConfigMapping]`

File: torch/ao/quantization/qconfig_mapping.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+            qconfig_mapping = QConfigMapping()
+                .set_module_name_regex("foo.*bar.*conv[0-9]+", qconfig1)
+                .set_module_name_regex("foo.*bar.*", qconfig2)
+                .set_module_name_regex("foo.*", qconfig3)
+
+        In this example, "foo.bar.conv0" would match qconfig1, "foo.bar.linear" would match qconfig2,
+        and "foo.baz.relu" would match qconfig3.
+
+        If the QConfig for an existing module name regex was already set, the new QConfig will override the
+        old one while preserving the order in which the regexes were originally registered.
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig_mapping.py:126]: how did we achieve this? looks like we just append the entry right now?

File: torch/ao/quantization/qconfig_mapping_utils.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+def convert_lists_to_ordered_dicts(qconfig_mapping: QConfigMapping):
+    """
+    Convert lists in a QConfigMapping to OrderedDicts.
+    """
+    qconfig_mapping._object_type_qconfig_dict = OrderedDict(
+        [(e.object_type, e.qconfig) for e in qconfig_mapping.object_type_qconfigs])
+    qconfig_mapping._module_name_qconfig_dict = OrderedDict(
+        [(e.module_name, e.qconfig) for e in qconfig_mapping.module_name_qconfigs])
+    qconfig_mapping._module_name_regex_qconfig_dict = OrderedDict(
+        [(e.module_name_regex, e.qconfig) for e in qconfig_mapping.module_name_regex_qconfigs])
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig_mapping_utils.py:108]: so these fields (e.g. _object_type_qconfig_dict) are filled only when we call this function, otherwise they would be empty, should we always update them while changing the QConfigMapping?

**Review by jerryzh168 — APPROVED** [review]
looks great!

File: torch/ao/quantization/qconfig_mapping.py (hunk 10 above)
Comment (andrewor14) [reply torch/ao/quantization/qconfig_mapping.py:126]: It's when we convert the lists to OrderedDicts. The constructor preserves the original key orders but uses the latest values for each key

File: torch/ao/quantization/qconfig_mapping_utils.py (hunk 11 above)
Comment (jerryzh168) [reply torch/ao/quantization/qconfig_mapping_utils.py:108]: not sure if this is exactly safe, how do we make sure we updates the internal dicts whenever there is an update? would it be cleaner if we don't make this internal?

Comment (jerryzh168) [reply torch/ao/quantization/qconfig_mapping_utils.py:108]: if we always regenerate when before we use it, I feel it might be cleaner that we just don't store this as a internal dict since when we store it, there will some confusion around how do we keep them in sync.

Comment (andrewor14) [reply torch/ao/quantization/qconfig_mapping_utils.py:108]: I think we do need to keep these fields however because today we pass around `QConfigMapping` everywhere expecting ordered dicts. What we could do is remove these fields from the constructor, but then we would need to suppress mypy warnings everywhere we access the ordered dicts.

Comment (andrewor14) [reply torch/ao/quantization/qconfig_mapping_utils.py:108]: I think we do need to keep these fields because today we pass around QConfigMapping everywhere expecting ordered dicts. What we could do is remove these fields from the constructor, but then we would need to suppress mypy warnings everywhere we access the ordered dicts.

**Review by albanD — COMMENTED** [review]
Looks good! Thanks for the fix!

File: torch/ao/quantization/qconfig_mapping_utils.py (hunk 11 above)
Comment (andrewor14) [reply torch/ao/quantization/qconfig_mapping_utils.py:108]: I think the semantics are fine if we always call `convert_lists_to_ordered_dicts(qconfig_mapping)` internally before each use. I also thought about freezing the lists after the populating the ordered dicts to prevent further changes, but I wasn't sure if the extra complexity is worth it.

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36597438).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36597438).

File: torch/ao/quantization/qconfig_mapping_utils.py (hunk 11 above)
Comment (jerryzh168) [reply torch/ao/quantization/qconfig_mapping_utils.py:108]: but calling `convert_lists_to_ordered_dicts` is not visible to QConfigMapping right? it's outside of the definition of QConfigMapping and nothing is guarantee that will be called, alternatively I feel we could also have a `build` method for QConfigMapping that will generate these internal dicts

Comment (andrewor14) [reply torch/ao/quantization/qconfig_mapping_utils.py:108]: Ok, I was actually just preserving the existing behavior in this PR (which actually even replaced the lists using the same dict fields for minimum type safety), and this was quite difficult to follow. I like option (1), where we just maintain ordered dicts. I think this is the simplest.

Comment (jerryzh168) [reply torch/ao/quantization/qconfig_mapping_utils.py:108]: yeah just use ordered_dicts directly sounds good to me as well

Comment (andrewor14) [reply torch/ao/quantization/qconfig_mapping_utils.py]: Sounds good. The new qconfig mapping uses ordered dicts directly. This also means we no longer need the dataclasses `QConfig*Entry`, which can't be used as dict keys because they are mutable.

File: torch/ao/quantization/qconfig_mapping.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        Set the QConfig for modules matching a combination of the given module name, object type,
+        and the index at which the module appears.
+
+        If the QConfig for an existing (module name, object type, index)  was already set, the new QConfig
+        will override the old one.
+        """
+        self.module_name_object_type_order_qconfigs[(module_name, object_type, index)] = qconfig
+        return self
+
+    def to_dict(self) -> Dict[str, Any]:
```
Comment (vkuzo) [inline torch/ao/quantization/qconfig_mapping.py:116]: should we add a TODO to remove this in the future?

Comment (vkuzo) [reply torch/ao/quantization/qconfig_mapping.py:116]: it seems reasonable to have it in this PR to break things into chunks instead of a huge PR which fixes everything, but in the future ideally the workflow code just uses the new object directly without having to convert to `qconfig_dict` again

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36597438).

File: torch/ao/quantization/qconfig_mapping.py (hunk 12 above)
Comment (andrewor14) [reply torch/ao/quantization/qconfig_mapping.py:116]: Well actually `to_dict` is not currently used anywhere and can be safely removed without widening the scope of the PR. I think you're thinking about `from_dict`? I can add a TODO to remove that method

Comment (andrewor14) [reply torch/ao/quantization/qconfig_mapping.py:116]: Well actually `to_dict` is not currently used anywhere and can be safely removed without widening the scope of the PR (internally we always handle `QConfigMapping` directly). I think you're referring to `from_dict`? I can add a TODO to remove that method

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36597438).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36597438).

**Comment by andrewor14** [comment]
Closed in favor of #78452.
